### PR TITLE
Infer value of date objects fixes #1952

### LINKED ISSUE #1952: Gatsby Build Doesn't Work Without Cache
Here is the source for my work-in-progress site: https://github.com/knpwrs/knpw.rs

When I run `gatsby build`, I get the following output:

```
❯ npm run build

> knpw.rs@ build /Users/knpwrs/Workspace/github.com/knpwrs/knpw.rs
> gatsby build

success delete html files from previous builds — 0.005 s
success open and validate gatsby-config.js — 0.004 s
success copy gatsby files — 0.019 s
success source and transform nodes — 0.094 s
success building schema — 0.283 s
success createLayouts — 0.007 s
error Plugin default-site-plugin returned an error


  Error: 

success createPages — 0.024 s
success createPagesStatefully — 0.025 s
GraphQL Error Unknown field `date` on type `frontmatter`

  file: /Users/knpwrs/Workspace/github.com/knpwrs/knpw.rs/src/pages/index.jsx

   1 |
   2 |   query IndexQuery {
   3 |     allMarkdownRemark(
   4 |       sort: { order: DESC, fields: [frontmatter___date] }
   5 |     ) {
   6 |       edges {
   7 |         node {
   8 |           id
   9 |           frontmatter {
  10 |             title
> 11 |             date(formatString: "MMMM DD, YYYY")
     |             ^
  12 |             path
  13 |             tags
  14 |           }
  15 |         }
  16 |       }
  17 |     }
  18 |   }
  19 |

success extract queries from components — 0.155 s
success run graphql queries — 0.002 s
success write out page data — 0.003 s
success update schema — 0.083 s

info bootstrap finished - 1.829 s

success Generating CSS — 4.858 s
success Compiling production bundle.js — 12.092 s

error Generating static HTML for pages failed
```

When I run `gatsby develop` I'll get a `.cache` directory which will then make `gatsby build` work. If I delete that cache directory then `gatsby build` goes back to the previous output. Is this intentional? Whats' going on?

### PR DESCRIPTION
**Pull request by KyleAMathews**

### CODE REVIEW

**Comment by KyleAMathews** [comment]
@jquense seem reasonable?

File: packages/gatsby/src/schema/data-tree-utils.js [hunk 1]
```diff
@@ -18,6 +18,8 @@ const areAllSameType = list =>
 const isEmptyObjectOrArray = (obj: any): boolean => {
   if (obj === INVALID_VALUE) {
     return true
+  } else if (_.isDate(obj)) {
+    return false
```
Comment (jquense) [inline packages/gatsby/src/schema/data-tree-utils.js]: does the below branch not catch this already?

File: packages/gatsby/src/schema/infer-graphql-type.js [hunk 2]
```diff
@@ -129,7 +129,7 @@ function inferGraphQLType({
         },
       },
       resolve(object, { fromNow, difference, formatString }) {
-        const date = object[fieldName]
+        const date = object[fieldName].toString()
```
Comment (jquense) [inline packages/gatsby/src/schema/infer-graphql-type.js]: does `toISOString()` make more sense as a default? its a bit less culture dependent

File: packages/gatsby/src/schema/data-tree-utils.js (hunk 1 above)
Comment (KyleAMathews) [reply packages/gatsby/src/schema/data-tree-utils.js]: No it's reported as an empty object

Comment (jquense) [reply packages/gatsby/src/schema/data-tree-utils.js]: weeiiiird

File: packages/gatsby/src/schema/infer-graphql-type.js (hunk 2 above)
Comment (KyleAMathews) [reply packages/gatsby/src/schema/infer-graphql-type.js]: Oh for sure!

Comment (KyleAMathews) [reply packages/gatsby/src/schema/infer-graphql-type.js]: Will fix later

Comment (KyleAMathews) [reply packages/gatsby/src/schema/infer-graphql-type.js]: Needed to do JSON.parse(JSON.stringify(date)) as date could just be a string so this seemed simplest. The internets say that JSON.stringify uses Date.toISOString under the hood anyways.

**Comment by KyleAMathews** [comment]
Deploy preview failed.

Built with commit 4def48c43c9ac2ed4dc9ac08c0898ba13f5ddb5c

https://app.netlify.com/sites/using-glamor/deploys/59a63d61cf321c54a9547048

**Comment by KyleAMathews** [comment]
Yeah! Travis is back to green!

**Pull request merged**